Unify devservices datatabase, username and password and allow configuration

newur opened the pull request:
I wanted to inspect the Postgres database started by devservices. While searching for the login credentials I noticed that different databases use different credentials and db names. I think it is nicer to use the same for all. Bonus, at least for Postgres the login via `psql` becomes shorter when the username and db name are equal

```psql -h localhost -p <randomPort> -U quarkus```

instead of

```psql -h localhost -p <randomPort> -U quarkus -d default```

I did not find the information how to connect to the devservices datasources in the docu. Would be worth to add it IMO. Any hints where I should put it?

Commit:
Unify Dev Services datasource names

gsmet reviewed (changes requested):
I think that's indeed a very good idea, thanks.

Only one small gripe: we avoid star imports in Quarkus. Any chance you could remove them?

gsmet commented:
Also please squash all the commits when done.

gsmet commented:
I force pushed a formatting fix to your branch.

newur commented:
DB2 really putted on a fight. Will the light prevail now or is it another round? ¯\_(ツ)_/¯

gsmet commented on extensions/datasource/deployment-spi/src/main/java/io/quarkus/datasource/deployment/spi/DatabaseDefaultSetupConfig.java:
```diff
@@ -0,0 +1,11 @@
+    // mssql container enforces a 'strong' password with min 8 chars, upper/lowercase, number or special char
+    // db2 derives the instance name from this. Enforces max 8 chars, only lowercase, picky on special chars
+    String DEFAULT_DATABASE_PASSWORD = "quarku_1";
+    String DEFAULT_DATABASE_NAME = "quarkus";
```
We don't use the constant in interface pattern. This needs to be a final class with a private constructor and public fields.

gsmet commented on extensions/devservices/mariadb/src/main/java/io/quarkus/devservices/mariadb/deployment/MariaDBDevServicesProcessor.java:
```diff
@@ -39,9 +43,9 @@ public RunningDevServicesDatasource startDatabase(Optional<String> username, Opt
-                container.withPassword(password.orElse("quarkus"))
-                        .withUsername(username.orElse("quarkus"))
-                        .withDatabaseName(datasourceName.orElse("default"));
+                container.withPassword(password.orElse(DEFAULT_DATABASE_USERNAME))
+                        .withUsername(username.orElse(DEFAULT_DATABASE_PASSWORD))
+                        .withDatabaseName(datasourceName.orElse(DEFAULT_DATABASE_NAME));
```
You mixed passwords and usernames up. That's probably why you had issues. Please check all of them and put the username first as it makes more sense.

gsmet commented on extensions/jdbc/jdbc-postgresql/deployment/src/test/java/io/quarkus/jdbc/postgresql/deployment/DevServicesPostgresqlDatasourceTestCase.java:
```diff
@@ -48,7 +48,7 @@ public void testDatasource() throws Exception {
         }
         assertTrue(configuration.connectionFactoryConfiguration().jdbcUrl().contains("jdbc:postgresql:"));
-        assertEquals("quarkus", configuration.connectionFactoryConfiguration().principal().getName());
+        assertEquals("quarku_1", configuration.connectionFactoryConfiguration().principal().getName());
```
Can't you use the constants? BTW, this is a consequence of the mix you made between username and password. This should be the username here.

newur replied:
This would require adding a dependency.

gsmet commented on extensions/datasource/deployment-spi/src/main/java/io/quarkus/datasource/deployment/spi/DatabaseDefaultSetupConfig.java:
```diff
@@ -0,0 +1,11 @@
+
+    // mssql container enforces a 'strong' password with min 8 chars, upper/lowercase, number or special char
+    // db2 derives the instance name from this. Enforces max 8 chars, only lowercase, picky on special chars
+    String DEFAULT_DATABASE_PASSWORD = "quarku_1";
```
Please put the username first, it makes more sense.

gsmet commented on extensions/datasource/deployment-spi/src/main/java/io/quarkus/datasource/deployment/spi/DatabaseDefaultSetupConfig.java:
```diff
@@ -0,0 +1,11 @@
+
+    String DEFAULT_DATABASE_USERNAME = "quarkus";
+
+    // mssql container enforces a 'strong' password with min 8 chars, upper/lowercase, number or special char
+    // db2 derives the instance name from this. Enforces max 8 chars, only lowercase, picky on special chars
```
There's a chance this comment is incorrect and the problem is due to the big mix up between usernames and passwords.

gsmet reviewed (changes requested):
Please see my comments. There is a big mix up between username and password in the patch.

newur commented:
Not sure why this test fails. Works here with JDK 11 when I run it via `mvn test`.

GavinRay97 commented:
Would this be a breaking change, for anyone who had manual connection strings specifying the database in their tests before?

newur commented:
> Would this be a breaking change, for anyone who had manual connection strings specifying the database in their tests before?

I would not think so. The whole point of the dev services is that your app does not specify any connection details.

gsmet commented on extensions/devservices/h2/src/main/java/io/quarkus/devservices/h2/deployment/H2DevServicesProcessor.java:
```diff
@@ -49,12 +53,12 @@ public RunningDevServicesDatasource startDatabase(Optional<String> username, Opt
-                            "sa",
-                            "sa",
+                            DEFAULT_DATABASE_USERNAME,
+                            DEFAULT_DATABASE_PASSWORD,
```
Just checking but we have the full permissions when doing that and not using `sa`? I thought it had a special meaning? Given I don't see us defining permissions for the `quarkus` user, I'm wondering how this works?

newur replied:
> By default, if the database specified in the [embedded](https://h2database.com/html/features.html#database_url) URL does not yet exist, a new (empty) database is created automatically. The user that created the database automatically becomes the administrator of this database.

From H2 docu https://h2database.com/html/tutorial.html#creating_new_databases
`sa` is the abriviation for `System Administrator`, but seems to have no special permissions. It is just some default IMHO.

gsmet commented on extensions/devservices/mssql/src/main/java/io/quarkus/devservices/mssql/deployment/MSSQLDevServicesProcessor.java:
```diff
@@ -35,7 +37,7 @@ public RunningDevServicesDatasource startDatabase(Optional<String> username, Opt
                         !devServicesSharedNetworkBuildItem.isEmpty());
                 startupTimeout.ifPresent(container::withStartupTimeout);
-                container.withPassword(password.orElse("Quarkuspassword1"))
+                container.withPassword(password.orElse(DEFAULT_DATABASE_PASSWORD))
```
We don't define a username, here?

newur replied:
I was also surprised, but did not dig deeper into this one.

gsmet commented on extensions/datasource/deployment-spi/src/main/java/io/quarkus/datasource/deployment/spi/DatabaseDefaultSetupConfig.java:
```diff
@@ -0,0 +1,13 @@
+    public static final String DEFAULT_DATABASE_USERNAME = "quarkus";
+
+    // mssql container enforces a 'strong' password with min 8 chars, upper/lowercase, number or special char
+    public static final String DEFAULT_DATABASE_PASSWORD = "quarkus1!";
```
I think I would rather use `quarkus` for the password and have a `STRONG_DEFAULT_DATABASE_PASSWORD` for databases that requires it so only MSSQL for now.

Rationale is that in some cases you want to connect to the database while dev mode is running. And it's better to have a simple password if we can.

gsmet reviewed (commented):
I added some suggestions and questions. Could you have a look?

I want to make sure we don't do any mistakes here.

(I think we're nearly there)

gastaldi commented:
This PR needs a rebase since the `H2DevServicesProcessor` class changed recently

gsmet commented:
I'm working on this as I want to make other changes in this area and thought it would be better to build on this one.

gsmet committed:
Some documentation improvements around Database Dev Services unification

gsmet committed:
Allow easily overriding database, username and password for Dev Services

Not all database containers support everything but it's implemented for
all everything that is supported.

gsmet reviewed (approved):
I made a few adjustments for the doc and also added a commit to be able to easily override database names, usernames and passwords.

It's not obvious from the GitHub UI but authorship of the first commit has been preserved: `Author: newur <newur@mailbox.org>`. My guess is that there's a mismatch between email used in commit and emails declared on GitHub.

gastaldi commented on extensions/devservices/mssql/src/main/java/io/quarkus/devservices/mssql/deployment/MSSQLDevServicesProcessor.java:
```diff
@@ -1,5 +1,7 @@
 package io.quarkus.devservices.mssql.deployment;
 
+import static io.quarkus.datasource.deployment.spi.DatabaseDefaultSetupConfig.DEFAULT_DATABASE_PASSWORD;
```
```suggestion
import static io.quarkus.datasource.deployment.spi.DatabaseDefaultSetupConfig.DEFAULT_DATABASE_STRONG_PASSWORD;
```

gastaldi reviewed (commented):
CI failures seem related, probably because it needs to use the `DEFAULT_DATABASE_STRONG_PASSWORD`. 

BTW I think the `DEFAULT_DATABASE_STRONG_PASSWORD` should be used in all databases instead of having just for MSSQL

gastaldi reviewed (commented):
CI failures seem related, probably because it needs to use the `DEFAULT_DATABASE_STRONG_PASSWORD`. 

BTW I think the `DEFAULT_DATABASE_STRONG_PASSWORD` should be used in all databases instead of having just for MSSQL

gsmet commented:
I merged it, it's a step in the right direction. Thanks @newur for starting this work!
